Source/CombatExtended/Utilities/Tracking/ThingsTracker.cs

```diff
@@ -0,0 +1,208 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using RimWorld;
+using UnityEngine;
+using Verse;
+
+namespace CombatExtended.Utilities
+{
+    public class ThingsTracker : MapComponent
+    {
+        private static bool[] validDefs;
+
+        private static Map[] maps = new Map[20];
+        private static ThingsTracker[] comps = new ThingsTracker[20];
+
+        public static ThingsTracker GetTracker(Map map)
+        {
+            var index = map.Index;
+            if (index >= 0 && index < comps.Length)
+            {
+                if (maps[index] == map)
+                    return comps[index];
+                maps[index] = map;
+                return comps[index] = map.GetComponent<ThingsTracker>();
+            }
+            return map.GetComponent<ThingsTracker>();
+        }
```

Comment:
We can utilise the overload `FinalizeInit` provided by MapComponent in order to grow this array as needed. See something similar to
```csharp
	public override void FinalizeInit()
	{
		if (map.Index > cache.Length)
		{
			Array.Resize(ref cache, map.Index + GROWTH_FACTOR);
		}
		cache[map.Index] = this;
	}

	public static FridgeCache GetTracker(int mapIndex /* could be Map as well */)
	{
                if(mapIndex < 0) throw 
		return cache[mapIndex];
	}
	```

Reply:
The index is recyclable. Maps recycle indexes all the time.

Reply:
ok will do